control/discovery.py

```diff
@@ -0,0 +1,1127 @@
+    ICREQ = 0x0
+    ICRESP = 0x1
+    H2C_TERM = 0x2
+    C2H_TERM = 0x3
+    CMD = 0x4
+    RSP = 0x5
+    H2C_DATA = 0x6
+    C2H_DATA = 0x7
+    TCP_R2T = 0x9
+
+# NVMe tcp opcode
+class NVME_TCP_OPC(enum.IntFlag):
+    DELETE_SQ = 0x0
+    CREATE_SQ = 0x1
+    GET_LOG_PAGE = 0x2
+    DELETE_CQ = 0x4
+    CREATE_CQ = 0x5
+    IDENTIFY = 0x6
+    ABORT = 0x8
+    SET_FEATURES = 0x9
+    GET_FEATURES = 0xa
+    ASYNC_EVE_REQ = 0xc
+    NS_MGMT = 0xd
+    FW_COMMIT = 0x10
+    FW_IMG_DOWNLOAD = 0x11
+    NS_ATTACH = 0x15
+    KEEP_ALIVE = 0x18
+    FABRIC_TYPE = 0x7F
+
+# NVMe tcp fabric command type
+class NVME_TCP_FCTYPE(enum.IntFlag):
+    PROP_SET = 0x0
+    CONNECT = 0x1
+    PROP_GET = 0x4
+    AUTH_SEND = 0x5
+    AUTH_RECV = 0x6
+    DISCONNECT = 0x8
+
+# NVMe controller register space offsets
+class NVME_CTL(enum.IntFlag):
+    CAPABILITIES = 0x0
+    VERSION = 0x08
+    CONFIGURATION = 0x14
+    STATUS = 0x1c
+
+
+# NVM subsystem types
+class NVMF_SUBTYPE(enum.IntFlag):
+    # Discovery type for NVM subsystem
+    DISCOVERY = 0x1
+    # NVMe type for NVM subsystem
+    NVME = 0x2
+
+# NVMe over Fabrics transport types
+transport_types = {
+    "RDMA": 0x1,
+    "FC": 0x2,
+    "TCP": 0x3,
+    "INTRA_HOST": 0xfe
+}
+
+# Address family types
+adrfam_types = {
+    "ipv4": 0x1,
+    "ipv6": 0x2,
+    "ib": 0x3,
+    "fc": 0x4,
+    "intra_host": 0xfe
+}
+
+# Transport requirement, secure channel requirements
+# Connections shall be made over a fabric secure channel
+class NVMF_TREQ_SECURE_CHANNEL(enum.IntFlag):
+    NOT_SPECIFIED = 0x0
+    REQUIRED = 0x1
+    NOT_REQUIRED = 0x2
+
+# maximum number of connections
+MAX_CONNECTION = 10240
+
+# NVMe tcp package length, refer: MTU = 1500 bytes
+NVME_TCP_PDU_UNIT = 1024
+
+# Max SQ head pointer
+SQ_HEAD_MAX = 128
+
+lock = threading.Lock()
+
+# Global controller id
+GLOBAL_CNLID = 0x1
+
+# Global generation counter
+GLOBAL_GEN_CNT = 0x1
+
+class Connection(Structure):
+    """Data used multiple times in each connection."""
+
+    _fields_ = [
+        ("nvmeof_connect_data_hostid", c_ubyte * 16),
+        ("nvmeof_connect_data_cntlid", c_uint16),
+        ("nvmeof_connect_data_subnqn", c_ubyte * 256),
+        ("nvmeof_connect_data_hostnqn", c_ubyte * 256),
+        ("sq_head_ptr", c_uint16),
+        ("unsent_log_page_len", c_uint32),
+        ("property_data", c_ubyte * 8),
+        ("property_set", c_bool),
+        ("shutdown_now", c_bool),
+        ("cnlid", c_uint16),
+        ("gen_cnt", c_uint16),
+        ("recv_async", c_bool),
+        ("async_cmd_id", c_uint16),
+        ("keep_alive_time", c_float),
+        ("keep_alive_timeout", c_uint32),
+    ]
+
+    def __init__(self):
+        self.connection = 0
+        self.allow_listeners = []
+        self.log_page = bytearray()
+        self.recv_buffer = b''
+
+        self.unsent_log_page_len = 0
+        self.property_set = False
+        self.shutdown_now = False
+        self.recv_async = False
```

Comment:
Perhaps the `UUID` type is lowercase `uuid`? Also, I think `str` may not be able to replace `c_ubytes * 8/16/256`. Is it possible? Please help me take a closer look at this section. My `types` and `initialization` were poorly written. Especially for initializing `c_ubytes * 8/16/256`, I need to ensure that it has real storage, because if there are independent requests to access this part of data that have not been initialized, it may cause errors due to accessing subscripts..

Reply:
> Yup, sorry, my replies don't always compile :P (I abuse of pseudo-code).
> 
Thank you, just because I am not proficient in Python and asked carefully.

> Ok, my doubt here was whether that structure was actually used to cast/wrap packed binary data, or just an intermediate structure where data was stored from the packed binary data. If the former is the case, then we obviously need ctypes (or any other Python package with binary interface).

This data structure is not used for transmission,  just an intermediate structure and is used to store some cross request data and some connection universal data. For example, when get `property_data` from request 1, and then get request 2, the response should be based on `property_data`. It does not require strict formatting(no need `ctypes`), but requires actual storage. Because there may are unexpected requests, such as the server closing the connection and the client still re-connectting and resending some request skip initialization request, it is necessary to ensure that it has actual storage so that there are no errors during access. 

Perhaps `str` can be initialized to a format with actual storage, like this ?
```
@dataclass
class MyClass:
    my_str: str = field(default='\x00' * 16, init=False)
```